### PR TITLE
[WIP] feat: support elastic ep downscale in DP

### PR DESCRIPTION
**Pull request by intermezzi**
支持pymodel decode的EP缩容
设计文档：https://alidocs.dingtalk.com/i/nodes/lyQod3RxJKe9QjOMidyvXEpoWkb4Mw9r
整体进展：https://alidocs.dingtalk.com/i/nodes/qnYMoO1rWxrkmoj2I071OQvDJ47Z3je9

### CODE REVIEW

**Comment by LLLLKKKK** [comment]
🤖 AI Code Review — PR #681
**Head SHA:** `364e493a6801a75b0fcb9119abb6c03f51d1fe2a` | **Verdict:** P2

### Summary
[WIP] Elastic EP downscale support in DP mode. When ranks become inactive (via DeepEP mask buffer), MoE expert assignments are redistributed. Adds `ElasticEPManager` (C++), extends `ExpertBalancer` with downscale path, adds `PyWrappedModel::updateLayerWeight`, and extends the EPLB algorithm for fewer active ranks.

### Findings (Architecture-focused)

**P2: `get_deepep_wrapper_if_initialized` exception handling bug**
Accesses `DeepEPWrapper._instance.config` which raises `AttributeError` (not `RuntimeError`) when `_instance` is `None`. The `except RuntimeError` won't catch it. Should check `_instance is not None` first or also catch `AttributeError`.

**P2: `executeDownScale` blocks inference synchronously**
Synchronous disk I/O (`loadBalanceWeight`) and GPU copies (`updateBalanceWeight`) for every layer on the executor hot path. Acceptable for WIP, but production needs async loading like the existing EPLB path.

**P2: Inactive rank insertion in `rebalance_experts` may be incorrect with multiple inactive ranks**
The loop inserts zero-slices and shifts `log2phy` indices, but `idx` is from the original list while the list grows with each insert. Multiple consecutive inactive ranks may produce wrong positions.

**P2: Missing trailing newline in `ElasticEPManager.cc` and `.h`**

**Minor:** `enable_H2D` parameter naming is non-PEP8 (mixed case).

Full review: `.ci_analysis/pr_681/code_review/cr_v1.md`

**Comment by LLLLKKKK** [comment]
## AI Code Review — PR #681

### PR 概述

**Title:** `[WIP] feat: support elastic ep downscale in DP`
**Author:** intermezzi
**规模:** 38 files, +1386/-131

在纯 DP decode + DeepEP low latency 模式下支持弹性 EP 缩容。当部分 rank 下线时，系统检测 active ranks 变化，重新分配 expert 权重到存活 rank 上。

---

### Review 意见

#### P0 问题（阻塞）

1. **ElasticEPManager 每步 forward 获取 GIL 调用 Python**

   `updateElasticEPStats()` 在每次 `stepForward` 时都获取 GIL 并调用 Python `query_active_ranks()`，内部还有两次 `torch.cuda.synchronize()`。在高 QPS 下会显著影响推理延迟。

   建议：将 query 频率降低为每 N 步执行一次，或增加可配置的 `elastic_ep_check_interval` 参数。

2. **ElasticEPStats 成员 active_ranks_num_ 未初始化**

   `ElasticEPStats` 中 `active_ranks_num_`、`ep_rank_`、`ep_size_` 没有默认初始化值。当 `elastic_ep_manager_` 为 null 时，未初始化的 `ep_stats` 会被传递给 `expert_balancer_->stepForward()`，其中空 tensor `active_ranks_tensor_cpu_` 最终传入 Python 侧可能导致未定义行为。

   建议：给所有成员加默认值，确保空 tensor 在 Python 侧被正确处理。

3. **get_deepep_wrapper_if_initialized 访问未初始化 _instance.config 抛 AttributeError**

   ```python
   def get_deepep_wrapper_if_initialized():
       try:
           return DeepEPWrapper.get_instance(DeepEPWrapper._instance.config)
       except RuntimeError:
           return None
   ```
   当 `_instance` 为 None 时，`_instance.config` 抛 `AttributeError`，不会被 `except RuntimeError` 捕获，异常传播到 C++ 层。

   建议：先检查 `DeepEPWrapper.is_initialized()` 再访问 instance。

#### P1 问题（重要）

4. **executeDownScale 失败后状态不一致，不会重试**

   `last_active_ranks_num_` 在 `executeDownScale` 之前就已更新。如果 `executeDownScale` 抛异常，下一步 `is_downscale_` 为 false，downscale 不会重试，系统处于权重部分更新的不一致状态。

   建议：在 `executeDownScale` 成功后才更新 `last_active_ranks_num_`。

5. **ProcessManager elastic EP 模式下完全不处理进程死亡**

   `enable_elastic_ep=True` 时，子进程死亡后不做任何处理。如果所有进程都死了但仍有存活进程，主进程会永远 hang。force kill 超时也被跳过。

   建议：增加更精细的策略，至少在所有进程都死亡时应该退出。

6. **checkDownScale 每步 forward 打 LOG_INFO**

   两个分支都打 LOG_INFO，高 QPS 下产生大量日志。建议改为 `LOG_EVERY_N` 或仅在状态变化时打印。

7. **eplb.py downscale 路径 log2phy 偏移逻辑在多个不活跃 rank 时可能不正确**

   循环中修改 `phy2log_slices` 列表（insert）导致后续迭代的 idx 与原始列表不对齐。`log2phy` 的 `torch.where` 偏移也会累积错误。建议添加多个不活跃 rank 的单元测试。

8. **各 rank 独立查询 active_ranks 无同步保证**

   各 rank 独立调用 `query_active_ranks()`，没有 barrier 保证一致性。如果不同 rank 看到不同的 active ranks 状态，部分 rank 进入 `executeDownScale`（包含 allgather）而其他不进入，会导致 hang。

   建议：在 downscale 决策前增加显式的跨 rank 同步。

#### P2 问题（建议）

9. `eplb.py` 中 `torch.set_printoptions(threshold=10_000)` 是调试残留，影响全局设置
10. `ExpertBalancerPythonWrapper.h` 不必要包含 `PyWrappedModel.h`，应移到 `.cc`
11. NormalExecutor / MtpExecutor 中 elastic EP 调用代码完全重复，建议提取

#### 小问题
- `allgather_intput_ints` 拼写错误（应为 `input`）
- `ElasticEPManager.h/cc` 文件末尾缺少换行符
- `moe_group_args.py` 大量引号风格变更应与功能改动分离
- `deepep_test.py` 中 `print(f"gather_tensor: ...)` 是调试残留

---

### 整体评价

架构方向合理，但存在 3 个 P0 和 4 个 P1 问题。核心风险：(1) 每步 forward 获取 GIL + GPU sync 的性能影响；(2) 分布式一致性缺少跨 rank 同步保证；(3) 异常恢复路径不完整。建议 WIP 阶段优先解决这些问题。

**结论：存在阻塞/重要问题，不建议合入。**

**Comment by LLLLKKKK** [comment]
## AI Code Review - PR #681

**Status:** BLOCKING

**Summary:** P0/0 · P1/4 · P2/6 · P3/3

### Blocking Issues

#### P1
- **get_deepep_wrapper_if_initialized 未初始化时抛 AttributeError 而非返回 None** @ `rtp_llm/models_py/distributed/deepep_wrapper.py:578`
  - 建议：改为先检查 `if cls._instance is None: return None`，或将 except 改为 `except (RuntimeError, AttributeError): return None`。
- **ElasticEPStats 的 ep_rank_、ep_size_、active_ranks_num_ 无默认初始化，条件路径下传播未定义值** @ `rtp_llm/cpp/models/elastic_ep_manager/ElasticEPManager.h:4`
  - 建议：给所有 POD 成员加默认值：`int ep_rank_ = 0; int ep_size_ = 0; int active_ranks_num_ = 0;`。
- **rebalance_experts 中 scatter_ 的 arange 范围变更可能影响非缩容路径** @ `rtp_llm/eplb/eplb.py:261`
  - 建议：在非缩容路径（num_active_ranks == num_gpus）保持原始 `torch.arange(num_replicas, ...)` 行为，仅在缩容路径使用新计算方式。
- **ProcessManager 启用 elastic EP 后进程死亡不做任何处理，可能导致僵尸状态** @ `rtp_llm/utils/process_manager.py:143`
  - 建议：添加 expected_dead_count 或 allowed_dead_pids 机制，而非完全禁用进程死亡处理。至少保留 force kill 超时兜底防止无限 hang，并在跳过 terminate 时仍打印日志。

### Non-blocking Suggestions

#### P2
- **ElasticEPManager::updateElasticEPStats 每步 forward 获取 GIL + GPU sync，影响 decode 延迟** @ `rtp_llm/cpp/models/elastic_ep_manager/ElasticEPManager.cc:11`
  - 建议：1) 缓存 py::module 和 deepep_wrapper 对象到成员变量避免每步 import；2) 降低查询频率（如每 N 步查询一次）；3) query_active_ranks 返回 CPU tensor 后在 C++ 侧处理避免 .item() 调用。
- **syncUpdateWeights 异常被吞掉，缩容同步失败时静默继续** @ `rtp_llm/cpp/models/eplb/ExpertBalancer.cc:482`
  - 建议：同步失败时应抛出异常或返回错误状态，让调用方 executeDownScale 能够感知并处理。
- **executeDownScale 未区分活跃/非活跃 rank，非活跃 rank 也执行完整权重加载** @ `rtp_llm/cpp/models/eplb/ExpertBalancer.cc:218`
  - 建议：在 executeDownScale 入口检查当前 rank 是否活跃，非活跃 rank 可能只需参与 syncUpdateWeights 同步而无需加载权重。
- **checkDownScale 每次调用都打印 INFO 日志，高频调用时产生大量日志** @ `rtp_llm/cpp/models/eplb/ExpertBalancer.cc:197`
  - 建议：改用 RTP_LLM_LOG_DEBUG 或添加频率控制（如仅在状态变化时打印）。
- **ElasticEPManager::updateElasticEPStats catch 块中重复获取 GIL** @ `rtp_llm/cpp/models/elastic_ep_manager/ElasticEPManager.cc:51`
  - 建议：移除 catch 块中的 `py::gil_scoped_acquire acquire;`，外层 acquire 的 scope 已覆盖 catch 块。
- **syncUpdateWeights 中 CPU 标量写入 CUDA tensor 触发隐式同步** @ `rtp_llm/cpp/models/eplb/ExpertBalancer.cc:481`
  - 建议：在 CPU 上构造 tensor 后一次性 .cuda()：`auto t = torch::zeros({ep_size_}, torch::kInt32); t[ep_rank_] = 1; auto allgather_tensor = t.cuda();`。

#### P3
- **allgather_intput_ints 变量名拼写错误** @ `rtp_llm/models_py/distributed/deepep_wrapper.py:499`
  - 建议：重命名为 `allgather_input_ints`。
- **rebalance_experts 中 torch.set_printoptions 是调试残留** @ `rtp_llm/eplb/eplb.py:242`
  - 建议：移除 `torch.set_printoptions(threshold=10_000)` 调试代码。
- **ElasticEPManager.h 和 ElasticEPManager.cc 文件末尾缺少换行符** @ `rtp_llm/cpp/models/elastic_ep_manager/ElasticEPManager.h:25`
  - 建议：在文件末尾添加换行符。

<details>
<summary>Review Checklist: 6 pass / 10 fail</summary>

### General Principles Checklist

**Failed**
- [FAIL] 6.1 Software Engineering: 契约一致性：get_deepep_wrapper_if_initialized 的 docstring 声明返回 None 但实际可能抛异常  Linked issue: get_deepep_wrapper_if_initialized 未初始化时抛 AttributeError 而非返回 None
- [FAIL] 6.1 Architecture: 状态不变量：ElasticEPStats 在所有路径下成员均已初始化  Linked issue: ElasticEPStats 的 ep_rank_、ep_size_、active_ranks_num_ 无默认初始化，条件路径下传播未定义值
- [FAIL] 6.1 Architecture: 错误语义：fail-fast/retry/fallback/silent 行为明确  Linked issue: syncUpdateWeights 异常被吞掉，缩容同步失败时静默继续
- [FAIL] 6.1 Quality: 无调试代码残留  Linked issue: rebalance_experts 中 torch.set_printoptions 是调试残留

**Passed**
- [PASS] 6.1 Software Engineering: SRP：ElasticEPManager 职责单一，仅负责缩容状态检测，与 ExpertBalancer 解耦
- [PASS] 6.1 Tests: 新功能有对应的单元测试和集成测试
- [PASS] 6.1 Quality: 逻辑变更不混入无关格式化

### RTP-LLM Checklist

**Failed**
- [FAIL] B 正确性与逻辑: C++ struct POD 成员有默认初始化（pattern 1.71）  Linked issue: ElasticEPStats 的 ep_rank_、ep_size_、active_ranks_num_ 无默认初始化，条件路径下传播未定义值
- [FAIL] B 正确性与逻辑: 接口返回类型/行为变更有兼容处理（pattern 1.55）  Linked issue: rebalance_experts 中 scatter_ 的 arange 范围变更可能影响非缩容路径
- [FAIL] D 性能: per-forward C++→Python 调用引入 GIL + GPU sync（pattern 1.69）  Linked issue: ElasticEPManager::updateElasticEPStats 每步 forward 获取 GIL + GPU sync，影响 decode 延迟
- [FAIL] D 性能: CPU 标量写入 CUDA tensor 触发隐式同步（pattern 1.80）  Linked issue: syncUpdateWeights 中 CPU 标量写入 CUDA tensor 触发隐式同步
- [FAIL] G 跨语言框架陷阱: Python 函数调用异常类型与 C++ catch 匹配（pattern 1.34 调试残留）  Linked issue: get_deepep_wrapper_if_initialized 未初始化时抛 AttributeError 而非返回 None

**Passed**
- [PASS] A 兼容性与配置: MoeConfig 新增 enable_elastic_ep 字段有 C++ 默认值，pybind pickle 兼容

### Python Static-First Checklist

**Failed**
- [FAIL] P.B 错误处理: 异常捕获类型与实际抛出类型匹配（P.13）  Linked issue: get_deepep_wrapper_if_initialized 未初始化时抛 AttributeError 而非返回 None

**Passed**
- [PASS] P.F 语言陷阱: rebalance_experts active_ranks 参数类型标注正确（P.47）
- [PASS] P.H 类型标注: 函数返回类型标注准确（P.47/P.49）

</details>

### Strengths
- 弹性 EP 缩容功能设计合理，ElasticEPManager 独立管理缩容状态检测，与现有 EPLB 流程解耦，职责清晰
- CUDA kernel 层面 log_stats nullptr 保护（experts_stats_kernels.cu）使 pymodel 路径可复用 expert balance kernel，避免空指针崩溃
- 新增完整的 SelectTopkOp convert_logical_to_physical_experts 单元测试，覆盖多种参数组合和异常 dtype 场景
- DeepEP low latency downscale 测试覆盖 dispatch/combine/clean/allgather 四种 API 的故障模拟，测试设计严谨
